docs/source/blueprints/BP003.rst

```diff
@@ -0,0 +1,366 @@
+Under the hood, it means that the N(ext) Runner is not well integrated
+enough to an Avocado Job.
+
+A partial implementation of the N(ext) Runner integration to an
+Avocado Job is available at ``avocado/plugins/runner_nrunner.py`` but
+it has many limitations.  This blueprint describes the architecture of
+a scheduler that will power this Job compatible N(ext) Runner
+implementation.
+
+Motivation
+**********
+
+Describe the architecture of an task scheduler that will power the Job
+compatible N(ext) Runner implementation, allowing the N(ext) Runner
+first supported version to achieve feature completeness, and replace
+the current runner soon after the next LTS version is released.
+
+Specification
+*************
+
+Task execution coordination goals
+---------------------------------
+
+For an Avocado job to be useful with the N(ext) Runner architecture,
+tasks must be executed. Differently then the current runner, the
+N(ext) Runner architecture allows those to be executed in a much more
+decoupled way. This characteristic will be maintained, but it needs to
+be adapted into the current Job execution.
+
+From a high-level view, the nrunner plugin needs to:
+
+1. Break apart from the "one at a time" Task execution model that it currently
+   employs;
+
+2. Check if a Tasks can be executed, that is, if its requirements can be
+   fulfilled (the most basic requirement for a task is a matching runner);
+
+3. Prepare for the execution of a task, such as the fulfillment of extra task
+   requirements. The requirements resolver is one, if not the only way,
+   component that should be given a chance to act here;
+
+4. Executes a task in a prepared environment;
+
+5. Monitor the execution of a task (from an external PoV);
+
+6. Collect the status messages that tasks will send;
+
+  a. Forward the status messages to the appropriate job components, such as the
+  result plugins.
+
+  b. Depending on the content of messages, such as the ones containing "status:
+  started" or "status: finished", interfere in the Task execution status, and
+  consequently, in the Job execution status.
+
+7. Verify, warn the user, and attempt to clean up stray tasks. This may be, for
+   instance, necessary if a Task on a container seems to be stuck, and the
+   container can not be destroyed. The same applies to the process in some time
+   of uninterruptible periods of sleep.
+
+Parallelization
+---------------
+
+Because the N(ext) Runner features allow for parallel execution of tasks, all
+other aspects of task execution coordination (fulfilling requirements,
+collecting results, etc.) should not block each other.
+
+There are a number of strategies for concurrent programming in Python these
+days, and the "avocado nrun" command currently makes use of asyncio to have
+coroutines that spawn tasks and collect results concurrently (in a preemptive
+cooperative model). The actual language or library features used is, IMO, less
+important than the end result.
```

Comment:
> As soon as it is extensible and flexible.

Sure.  That's the kind of premise that should "go without saying", and that should be flagged either during development or  code review. For instance: "this won't scale because of Python's GIL. Running it with 10 tasks and you'll hit a 70% performance degradation".

At the same time, it's very abstract.  How extensible does it have to be?  How flexible does it have to be?  How do you measure those?